pkg/kf/command: integration test is flaky

Occasionally, the integration test has the following error:

```
Error: waiting for container termination: watch cancelled
waiting for container termination: watch cancelled
done pushing app "integration-echo-1558728385857498000".
ensuring app has domain...
panic: push "integration-echo-1558728385857498000": exit status 1

goroutine 194 [running]:
github.com/GoogleCloudPlatform/kf/pkg/kf/testutil.PanicOnError.func1(0x27b9be0, 0xc0000ce300, 0xc0004926c0, 0xc000206b00, 0xc0004d65a0, 0x2b)
	/Users/poy/workspace/go/src/github.com/GoogleCloudPlatform/kf/pkg/kf/testutil/integration.go:253 +0x353
created by github.com/GoogleCloudPlatform/kf/pkg/kf/testutil.PanicOnError
	/Users/poy/workspace/go/src/github.com/GoogleCloudPlatform/kf/pkg/kf/testutil/integration.go:243 +0x7f
FAIL	github.com/GoogleCloudPlatform/kf/pkg/kf/commands/apps	119.222s
```

The test normally passes the second time around. This leads me to believe that there is some startup time somewhere (build?) that pushes the required amount of time longer than the test likes.

--- a/pkg/kf/commands/apps/integration_test.go
+++ b/pkg/kf/commands/apps/integration_test.go
@@ -22,36 +22,28 @@ import (
 	"net/http"
 	"path/filepath"
 	"strings"
+	"sync"
 	"testing"
 	"time"
 
 	. "github.com/GoogleCloudPlatform/kf/pkg/kf/testutil"
 )
 
-// TestIntegration_Doctor runs the doctor command. It ensures the cluster the
-// tests are running against is in good shape.
-func TestIntegration_Doctor(t *testing.T) {
-	t.Parallel()
-	RunKfTest(t, func(ctx context.Context, t *testing.T, kf *Kf) {
-		kf.Doctor(ctx)
-	})
-}
-
 // TestIntegration_Push pushes the echo app (via --built-in command), lists it
 // to ensure it can find a domain, uses the proxy command and then posts to
 // it. It finally deletes the app.
 func TestIntegration_Push(t *testing.T) {
-	t.Parallel()
+	checkClusterStatus(t)
 	RunKfTest(t, func(ctx context.Context, t *testing.T, kf *Kf) {
 		appName := fmt.Sprintf("integration-echo-%d", time.Now().UnixNano())
 
 		// Push an app and then clean it up. This pushes the echo app which
 		// replies with the same body that was posted.
-		kf.Push(ctx, appName, map[string]string{
-			"--built-in":           "",
-			"--path":               filepath.Join(RootDir(ctx, t), "./samples/apps/echo"),
-			"--container-registry": fmt.Sprintf("gcr.io/%s", GCPProjectID()),
-		})
+		kf.Push(ctx, appName,
+			"--built-in",
+			"--path", filepath.Join(RootDir(ctx, t), "./samples/apps/echo"),
+			"--container-registry", fmt.Sprintf("gcr.io/%s", GCPProjectID()),
+		)
 		defer kf.Delete(ctx, appName)
 
 		// List the apps and make sure we can find a domain.
@@ -86,17 +78,17 @@ func TestIntegration_Push(t *testing.T) {
 // TestIntegration_Delete pushes an app and then deletes it. It then makes
 // sure it is marked as "Deleting".
 func TestIntegration_Delete(t *testing.T) {
-	t.Parallel()
+	checkClusterStatus(t)
 	RunKfTest(t, func(ctx context.Context, t *testing.T, kf *Kf) {
 		appName := fmt.Sprintf("integration-echo-%d", time.Now().UnixNano())
 
 		// Push an app and then clean it up. This pushes the echo app which
 		// simplies replies with the same body that was posted.
-		kf.Push(ctx, appName, map[string]string{
-			"--built-in":           "",
-			"--path":               filepath.Join(RootDir(ctx, t), "./samples/apps/echo"),
-			"--container-registry": fmt.Sprintf("gcr.io/%s", GCPProjectID()),
-		})
+		kf.Push(ctx, appName,
+			"--built-in",
+			"--path", filepath.Join(RootDir(ctx, t), "./samples/apps/echo"),
+			"--container-registry", fmt.Sprintf("gcr.io/%s", GCPProjectID()),
+		)
 
 		// This is only in place for cleanup if the test fails.
 		defer kf.Delete(ctx, appName)
@@ -123,7 +115,7 @@ func TestIntegration_Delete(t *testing.T) {
 // pushing, and another via SetEnv. It then reads them via Env. It then unsets
 // one via Unset and reads them again via Env.
 func TestIntegration_Envs(t *testing.T) {
-	t.Parallel()
+	checkClusterStatus(t)
 	RunKfTest(t, func(ctx context.Context, t *testing.T, kf *Kf) {
 		appName := fmt.Sprintf("integration-envs-%d", time.Now().UnixNano())
 
@@ -196,13 +188,13 @@ func TestIntegration_Envs(t *testing.T) {
 		// Push an app and then clean it up. This pushes the envs app which
 		// returns the set environment variables via JSON. Set two environment
 		// variables (ENV1 and ENV2).
-		kf.Push(ctx, appName, map[string]string{
-			"--built-in":           "",
-			"--path":               filepath.Join(RootDir(ctx, t), "./samples/apps/envs"),
-			"--container-registry": fmt.Sprintf("gcr.io/%s", GCPProjectID()),
-			"--env":                "ENV1=VALUE1",
-			"--env=ENV2=VALUE2":    "",
-		})
+		kf.Push(ctx, appName,
+			"--built-in",
+			"--path", filepath.Join(RootDir(ctx, t), "./samples/apps/envs"),
+			"--container-registry", fmt.Sprintf("gcr.io/%s", GCPProjectID()),
+			"--env", "ENV1=VALUE1",
+			"--env=ENV2=VALUE2",
+		)
 
 		// This is only in place for cleanup if the test fails.
 		defer kf.Delete(ctx, appName)
@@ -226,5 +218,43 @@ func TestIntegration_Envs(t *testing.T) {
 		checkVars(map[string]string{
 			"ENV2": "OVERWRITE2", // Set on push and overwritten via set-env
 		}, []string{"ENV1"})
+	})
+}
+
+var checkOnce sync.Once
+
+func checkClusterStatus(t *testing.T) {
+	checkOnce.Do(func() {
+		testIntegration_Buildpacks(t)
+		testIntegration_Doctor(t)
+	})
+}
+
+// testIntegration_Buildpacks uploads the sample buildpacks and then lists
+// them. It ensures the ablity to upload and list buildpacks along with
+// preparing the cluster for the rest of the integration tests (which require
+// the buildpacks).
+func testIntegration_Buildpacks(t *testing.T) {
+	RunKfTest(t, func(ctx context.Context, t *testing.T, kf *Kf) {
+		kf.UploadBuildpacks(ctx,
+			"--path", filepath.Join(RootDir(ctx, t), "./samples/buildpacks"),
+			"--container-registry", fmt.Sprintf("gcr.io/%s", GCPProjectID()),
+			"--built-in",
+		)
+
+		buildpacks := kf.Buildpacks(ctx)
+		AssertContainsAll(t, strings.Join(buildpacks, "\n"), []string{
+			"io.buildpacks.samples.nodejs",
+			"io.buildpacks.samples.go",
+			"io.buildpacks.samples.java",
+		})
+	})
+}
+
+// testIntegration_Doctor runs the doctor command. It ensures the cluster the
+// tests are running against is in good shape.
+func testIntegration_Doctor(t *testing.T) {
+	RunKfTest(t, func(ctx context.Context, t *testing.T, kf *Kf) {
+		kf.Doctor(ctx)
 	})
 }

--- a/pkg/kf/testutil/integration.go
+++ b/pkg/kf/testutil/integration.go
@@ -382,7 +382,7 @@ func KF(t *testing.T, kf KfInvoker) *Kf {
 }
 
 // Push pushes an application.
-func (k *Kf) Push(ctx context.Context, appName string, flags map[string]string) {
+func (k *Kf) Push(ctx context.Context, appName string, extraArgs ...string) {
 	k.t.Helper()
 	Logf(k.t, "pushing app %q...", appName)
 	defer Logf(k.t, "done pushing app %q.", appName)
@@ -392,17 +392,8 @@ func (k *Kf) Push(ctx context.Context, appName string, flags map[string]string) 
 		appName,
 	}
 
-	for k, v := range flags {
-		if v == "" {
-			args = append(args, k)
-			continue
-		}
-
-		args = append(args, fmt.Sprintf("%s=%s", k, v))
-	}
-
 	output, errs := k.kf(ctx, k.t, KfTestConfig{
-		Args: args,
+		Args: append(args, extraArgs...),
 	})
 	PanicOnError(ctx, k.t, fmt.Sprintf("push %q", appName), errs)
 	StreamOutput(ctx, k.t, output)
@@ -570,4 +561,36 @@ func (k *Kf) Doctor(ctx context.Context) {
 	})
 	PanicOnError(ctx, k.t, "doctor", errs)
 	StreamOutput(ctx, k.t, output)
+}
+
+// UploadBuildpacks runs the upload-buildpacks command.
+func (k *Kf) UploadBuildpacks(ctx context.Context, extraArgs ...string) {
+	k.t.Helper()
+	Logf(k.t, "running upload-buildpacks...")
+	defer Logf(k.t, "done running upload-buildpacks.")
+
+	args := []string{
+		"upload-buildpacks",
+	}
+
+	output, errs := k.kf(ctx, k.t, KfTestConfig{
+		Args: append(args, extraArgs...),
+	})
+	PanicOnError(ctx, k.t, "upload-buildpacks", errs)
+	StreamOutput(ctx, k.t, output)
+}
+
+// Buildpacks runs the buildpacks command.
+func (k *Kf) Buildpacks(ctx context.Context) []string {
+	k.t.Helper()
+	Logf(k.t, "running buildpacks...")
+	defer Logf(k.t, "done running buildpacks.")
+	output, errs := k.kf(ctx, k.t, KfTestConfig{
+		Args: []string{
+			"buildpacks",
+			"--built-in",
+		},
+	})
+	PanicOnError(ctx, k.t, "buildpacks", errs)
+	return CombineOutputStr(ctx, k.t, output)
 }

